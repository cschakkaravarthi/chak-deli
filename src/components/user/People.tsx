import React, { FC } from 'react';
import PeopleModel from '../../shared/models/People.model';
import Col from 'react-bootstrap/Col';
import { IndividualPersonCard } from 'umgc_ui_library/lib/cards/individualPersonCard';
import get from 'lodash.get';
import { Link as Go } from 'react-router-dom';
import { getUserProfilePicture } from '../../utils/customHooks';

type Props = {
  people: PeopleModel;
};

const People: FC<Props> = props => {
  const { people } = props;

  const imageUrl = getUserProfilePicture(props.people.avatar, true);

  return (
    <Col xs="6" md="4" className="mb-4 office-people">
      <IndividualPersonCard
        imageUrl={imageUrl}
        fullName={people.fullName}
        title={get(people, 'title', '')}
        MainLink={(p: React.PropsWithChildren<{}>) => (
          <Go
            to={{
              pathname: '/user',
              search: `?upn=${people.email}`
            }}
          >
            {p.children}
          </Go>
        )}
      />
    </Col>
  );
};

export default People;
