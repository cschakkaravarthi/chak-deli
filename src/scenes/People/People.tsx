import React, { FC, useEffect } from 'react';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import { connect } from 'react-redux';
import { match } from 'react-router';
import { Card } from 'umgc_ui_library/lib';
import { Link } from 'react-router-dom';
import get from 'lodash.get';
import find from 'lodash.find';

// Components
import TemplatesTopBar from '../../components/templatesTopBar/templatesTopBar';

// Types
import { AppState } from '../../shared/types/genericTypes';
import { contentToDepartment, contentToLabel, VariousContentModel } from '../../shared/types/contentTypes';
import { DepartmentTemplate, ContactGroup } from '../../shared/models/Department.model';
import { LabelTemplate } from '../../shared/models/Label.model';
import { Page as PageModel } from '../../shared/models/Page.model';
import UserInfoModel from '../../shared/models/UserInfo.model';

// Actions
import { fetchUsersByEmail } from '../../shared/actions/usersActions';
import { fetchDepartmentById } from '../../shared/actions/departmentActions';
import { fetchLabelById } from '../../shared/actions/labelActions';

type Props = {
  match?: match;
  givenId?: number;
  type: string;
  content?: VariousContentModel;
  contacts?: UserInfoModel[];
  fetchLabelById?: (labelId: number) => Promise<void>;
  fetchDepartmentById?: (departmentId: number) => Promise<void>;
  fetchUsersByEmail?: (email: string) => void;
};

const mapState: any = (state: AppState, props: Props): Props => {
  const givenId = parseInt(get(props, 'match.params.id'), 10);
  const givenContent = props.type === 'department'
    ? state.departmentReducers.departments.content
    : state.labelReducers.labels;

  return {
    ...props,
    givenId,
    contacts: state.usersReducers.users,
    content: find(givenContent, c => c.drupal_id === givenId)
  };
};

const actionCreators = { fetchLabelById, fetchDepartmentById, fetchUsersByEmail };

const getPeoplePage = (): PageModel => {
  return {
    type: 'page',
    language: 'en',
    categories: [],
    category_ancestors: [],
    owner: [],
    drupal_id: 0,
    title: 'People'
  };
};

const getGroupContacts = (contactGroup: ContactGroup, allContacts: (UserInfoModel[])): UserInfoModel[] => {
  contactGroup.contacts = contactGroup.contacts.map(c => c.toLowerCase());
  return allContacts.filter(c => {
    if (c && c.email) {
      return contactGroup.contacts.includes(c.email.toLowerCase());
    }
    return false;
  });
};

export const People: FC<Props> = props => {
  const {
    givenId,
    content,
    contacts,
    fetchLabelById,
    fetchDepartmentById,
    fetchUsersByEmail
  } = props;

  useEffect(() => {
    if (givenId && fetchDepartmentById && props.type === 'department') {
      fetchDepartmentById(givenId);
    } else if (givenId && fetchLabelById && props.type === 'label') {
      fetchLabelById(givenId);
    }
  }, [givenId, fetchDepartmentById, fetchLabelById]);

  let flatContacts: string[] = [];
  let typedContent = null;

  if (content) {
    typedContent = props.type === 'department'
      ? contentToDepartment(content)
      : contentToLabel(content);

    flatContacts = typedContent && typedContent.contacts
      ? typedContent.contacts.reduce((acc: string[], val) => acc.concat(val.contacts), [])
      : [];
  }

  useEffect(() => {
    if (flatContacts.length) {
      fetchUsersByEmail && fetchUsersByEmail(flatContacts.join(','));
    }
  }, [content]);

  const getTemplateId = (): string => {
    const getTemplates = props.type === 'department'
      ? get(content, 'template', [] as DepartmentTemplate[])
      : get(content, 'template', [] as LabelTemplate[]);
    return getTemplates[0].drupal_id.toString();
  };

  const templateId = content && getTemplateId();
  const page = getPeoplePage();

  return givenId && content && typedContent && contacts && templateId ? (
    <div className='people pb-5'>
      <TemplatesTopBar
        givenId={givenId}
        active={-1}
        page={page}
        type={props.type}
      />
      <Container>
        {typedContent.contacts && typedContent.contacts.map(contactGroup => {
          const groupContacts = getGroupContacts(contactGroup, contacts);
          return (
            <Row key={contactGroup.group_label || 'Contacts'}>
              <Col md={12} className='py-md-2 py-2'>
                <Card
                  title={contactGroup.group_label || 'Contacts'}
                  topBorder='green'
                  variant='standard'
                  type='cardsInContainer'>
                  <Row>
                    {groupContacts &&
                      groupContacts.map(c => {
                        const officeNumber = find(c.phone, { type: 'desk' });
                        const mobileNumber = find(c.phone, { type: 'mobile' });
                        return (
                          <Col key={c.email} md={6} className='py-1 px-3'>
                            <Card
                              {...{
                                type: 'contactMe',
                                variant: 'profile',
                                title: get(c, 'fullName', ''),
                                office: get(officeNumber, 'number', ''),
                                mobile: get(mobileNumber, 'number', ''),
                                email: c.email,
                                contactPic: get(c, 'profilePicture.medium'),
                                isCurrentUser: false,
                                LinkToProfile: (p: React.PropsWithChildren<{}>) => (
                                  <Link to={`/profile?upn=${c.email}`}>
                                    {p.children}
                                  </Link>
                                )
                              }}
                            />
                          </Col>
                        );
                      })}
                  </Row>
                </Card>
              </Col>
            </Row>
          );
        })}
      </Container>
    </div>
  ) : null;
};

export default connect(mapState, actionCreators)(People);
