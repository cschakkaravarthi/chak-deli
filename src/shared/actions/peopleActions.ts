import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';

import ApiService from '../services/apiService';

import {
  CLEAR_PEOPLE,
  ClearPeopleAction,
  FETCH_MORE_PEOPLE,
  LoadMorePeopleAction
} from '../types/peopleTypes';

import PeopleModel from '../models/People.model';
import { ApplicationState } from '../reducers';

export const loadPeople = (people: PeopleModel[]): LoadMorePeopleAction => ({
  people,
  type: FETCH_MORE_PEOPLE
});

type ThunkResult<R> = ThunkAction<R, ApplicationState, undefined, AnyAction>;

export const fetchPeople = (
  sitecode: string,
  startsWith?: string,
  limit?: number
): ThunkResult<void> => {
  return async (
    dispatch: ThunkDispatch<{}, {}, AnyAction>,
    getState: () => ApplicationState
  ): Promise<void> => {
    const {
      peopleReducers: { pageCount }
    } = getState();

    dispatch(() => ({ type: FETCH_MORE_PEOPLE }));
    return ApiService.getPeople(
      people => dispatch(loadPeople(people)),
      sitecode,
      startsWith,
      pageCount,
      limit
    );
  };
};

export const clearPeople = (): ClearPeopleAction => ({ type: CLEAR_PEOPLE });
