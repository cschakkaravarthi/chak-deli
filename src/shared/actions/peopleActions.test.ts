import { loadPeople } from './peopleActions';
import { dummyPeople } from '../models/People.model';

describe('peopleActions', () => {
  it('creates a loadPeople action', () => {
    expect(loadPeople([dummyPeople])).toMatchSnapshot();
  });
});
