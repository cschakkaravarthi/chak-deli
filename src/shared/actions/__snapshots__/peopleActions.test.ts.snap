// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`peopleActions creates a loadPeople action 1`] = `
Object {
  "people": Array [
    Object {
      "avatar": Object {
        "photo": "",
      },
      "company": "n/a",
      "costCenter": "GER30/DE050075",
      "country": "Germany",
      "department": "Sales",
      "email": "Bastian.vonPfister@studiocanal.de",
      "employeeNumber": "",
      "employeeType": "3rd Party",
      "firstName": "Bastian",
      "fullName": "Bastian vonPfister",
      "lastName": "vonPfister",
      "location": "",
      "loginId": "Balasua2",
      "managerName": "Joerg Bromberg",
      "mobileNumber": "",
      "phoneNumber": "+49 30 810969 213",
      "photo": Object {
        "photo": "",
      },
      "prefLanguage": "",
      "siteCode": "EXTNL",
      "title": "Manager Business Analysis & Commercial P",
      "voipNumber": "",
    },
  ],
  "type": "FETCH_MORE_PEOPLE",
}
`;
